### PR TITLE
SCons: Review uses of CCFLAGS, CXXFLAGS and CPPFLAGS

### PR DESCRIPTION
**Pull request by akien-mga**
Many contributors (me included) did not fully understand what CCFLAGS,
CXXFLAGS and CPPFLAGS refer to exactly, and were thus not using them
in the way they are intended to be.

As per the SCons manual: https://www.scons.org/doc/HTML/scons-user/apa.html

- CCFLAGS: General options that are passed to the C and C++ compilers.
- CFLAGS: General options that are passed to the C compiler (C only;
  not C++).
- CXXFLAGS: General options that are passed to the C++ compiler. By
  default, this includes the value of $CCFLAGS, so that setting
  $CCFLAGS affects both C and C++ compilation.
- CPPFLAGS: User-specified C preprocessor options. These will be
  included in any command that uses the C preprocessor, including not
  just compilation of C and C++ source files [...], but also [...]
  Fortran [...] and [...] assembly language source file[s].

TL;DR: Compiler options go to CCFLAGS, unless they must be restricted
to either C (CFLAGS) or C++ (CXXFLAGS). Preprocessor defines go to
CPPFLAGS.

---

Note: The documentation also mentions `CPPDEFINES` as a portable way to set preprocessor options. Some of our SCons scripts already use it, and we could think to general its use and avoid the ambiguous `CPPFLAGS`.

Supersedes #28373 (changes included in this PR too).

### CODE REVIEW

**Comment by Xrayez** [comment]
From a user perspective regarding `CPPDEFINES` this is how I do it for my of my own modules:
https://github.com/Xrayez/godot-anl/blob/master/SCsub#L27-L29

Just wonder is there anything from this PR that could be improved for such use cases?

**Review by Xrayez — APPROVED** [review]
Tested on MinGW from linux with:
`scons platform=windows target=release_debug tools=yes debug_symbols=no -j4`

**Comment by hpvb** [comment]
LGTM

**Comment by akien-mga** [comment]
> From a user perspective regarding CPPDEFINES this is how I do it for my of my own modules:
https://github.com/Xrayez/godot-anl/blob/master/SCsub#L27-L29

I can add a `CPPFLAGS` SCons option similar to the `CCFLAGS`, `CXXFLAGS` and `LINKFLAGS` that we already support: https://github.com/godotengine/godot/pull/28396/commits/c2a669a9f03e249fb4fca96a2adf63256eb8dfa2#diff-cc8093822b396591a5530a463c87d048R280

For now I think `CPPFLAGS` is more user friendly to expose as option, so that you pass exactly what you want the target compiler to use without having to worry about SCons' `CPPDEFINES` syntax (e.g. as a user I wouldn't know intuitively wether `scons CPPDEFINES="DEFINE1 DEFINE2"` will end up as `gcc -DDEFINE1 -DDEFINE2` or `gcc -D"DEFINE1 DEFINE2"`).

**Comment by akien-mga** [comment]
Note for release maintainers: Would be good to cherry-pick for 3.1.x to avoid later conflicts when cherry-picking other buildsystem changes. But there's potential for breakage, so this should be given some time for testing in the master branch.

**Comment by akien-mga** [comment]
Probably won't cherry-pick for 3.1.
